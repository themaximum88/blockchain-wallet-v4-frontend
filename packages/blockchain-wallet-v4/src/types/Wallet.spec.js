import * as R from 'ramda'
import { Address, Wallet, AddressMap, serializer } from './index'
import * as crypto from '../walletCrypto/index'

const walletFixture = require('./__mocks__/wallet.v3')
const walletFixtureSecpass = require('./__mocks__/wallet.v3-secpass')
const secpass = 'secret'

describe('Wallet', () => {
  const wallet = Wallet.fromJS(walletFixture)
  const walletSecpass = Wallet.fromJS(walletFixtureSecpass)

  describe('toJS', () => {
    it('should return the correct object', () => {
      let result = Wallet.toJS(wallet)
      expect(result).toEqual(walletFixture)
    })
  })

  describe('selectors', () => {
    it('should throw if it does not receive a Wallet instance', () => {
      expect(() => Wallet.select(Wallet.guid, {})).toThrow(TypeError)
    })

    it('should select guid', () => {
      expect(wallet.guid).toEqual(walletFixture.guid)
      expect(Wallet.selectGuid(wallet)).toEqual(walletFixture.guid)
    })

    it('should select iterations', () => {
      expect(Wallet.selectIterations(wallet)).toEqual(
        walletFixture.options.pbkdf2_iterations
      )
    })

    it('should select addresses', () => {
      let addressesJS = R.compose(
        AddressMap.toJS,
        Wallet.selectAddresses
      )(wallet)
      expect(addressesJS).toEqual(walletFixture.keys)
    })

    it('should select if is double encrypted', () => {
      expect(Wallet.isDoubleEncrypted(wallet)).toEqual(false)
      expect(Wallet.isDoubleEncrypted(walletSecpass)).toEqual(true)
    })
  })

  describe('importLegacyAddress', () => {
    const n = walletFixture.keys.length
    const address = Address.fromJS({
      addr: '1CC3X2gu58d6wXUWMffpuzN9JAfTUWu4Kj',
      priv: '5Kb8kLf9zgWQnogidDA76MzPL6TsZZY36hWXMssSzNydYXYB9KF'
    })

    it('should add an unencrypted address', () => {
      Wallet.importLegacyAddress(
        wallet,
        address.priv,
        0,
        undefined,
        undefined,
        {}
      ).fork(
        failure => expect(failure).toEqual(undefined),
        withNewAddress => {
          let addresses = Wallet.selectAddresses(withNewAddress)
          expect(addresses.size).toEqual(n + 1)
          const newAddr = R.compose(
            AddressMap.selectAddress('1CC3X2gu58d6wXUWMffpuzN9JAfTUWu4Kj'),
            Wallet.selectAddresses
          )(withNewAddress)
          expect(newAddr.addr).toEqual(address.addr)
          expect(newAddr.priv).toEqual(
            'GibbqZhygNfhbfz4fb2vDg19Ym5v696w52iqZEQySHTw'
          )
        }
      )
    })

    it('should add a double encrypted address', () => {
      Wallet.importLegacyAddress(
        wallet,
        address.priv,
        0,
        'secret',
        undefined,
        {}
      ).fork(
        failure => expect(failure).toEqual(undefined),
        withNewAddress => {
          let addresses = Wallet.selectAddresses(withNewAddress)
          expect(addresses.size).toEqual(n + 1)
          const newAddr = R.compose(
            AddressMap.selectAddress('1CC3X2gu58d6wXUWMffpuzN9JAfTUWu4Kj'),
            Wallet.selectAddresses
          )(withNewAddress)
          expect(newAddr.addr).toEqual(address.addr)
          expect(newAddr.priv).toEqual(
            'GibbqZhygNfhbfz4fb2vDg19Ym5v696w52iqZEQySHTw'
          ) // << TODO this is not encrypted?
        }
      )
    })
  })

  describe('setAddressLabel', () => {
    it('should set a new address label', () => {
      let addr = '14mQxLtEagsS8gYsdWJbzthFFuPDqDgtxQ'
      let withNewLabel = Wallet.setLegacyAddressLabel(addr, 'new_label', wallet)
      expect(wallet.addresses.get(addr).label).toEqual('labeled_imported')
      expect(withNewLabel.addresses.get(addr).label).toEqual('new_label')
    })
  })

  describe('isValidSecondPwd', () => {
    it('should be valid for an unencrypted wallet', async () => {
      const mockSecurityModule = {
        computeSecondPasswordHash: async () => walletSecpass.dpasswordhash
      }

      let isValid = await Wallet.isValidSecondPwd(
        mockSecurityModule,
        null,
        wallet
      )

      expect(isValid).toEqual(true)
    })

    it('should detect a valid second password', async () => {
      const securityModule = {
        computeSecondPasswordHash: async () => walletSecpass.dpasswordhash
      }

      let isValid = await Wallet.isValidSecondPwd(
        securityModule,
        secpass,
        walletSecpass
      )

      expect(isValid).toEqual(true)
    })

    it('should detect an invalid second password', async () => {
      const securityModule = {
        computeSecondPasswordHash: async () => `wrong hash`
      }

      let secpass = 'wrong_secpass'

      let isValid = await Wallet.isValidSecondPwd(
        securityModule,
        secpass,
        walletSecpass
      )

      expect(isValid).toEqual(false)
    })
  })

  describe('encrypt', () => {
    it('should encrypt', done => {
      const securityModule = {
        computeSecondPasswordHash: async ({ iterations, password }) =>
          `hash of ${JSON.stringify({ iterations, password })}`,

        encryptWithSecondPassword: async (
          { iterations, password },
          plaintext
        ) =>
          `encrypted with ${JSON.stringify({
            iterations,
            password,
            plaintext
          })}`
      }

      Wallet.encrypt(securityModule, 'secret', wallet).fork(done, encrypted => {
        expect(encrypted).toMatchSnapshot()
        done()
      })
    })
  })

  describe('decrypt', () => {
    it('should decrypt', done => {
      const securityModule = {
        computeSecondPasswordHash: async () => walletSecpass.dpasswordhash,

        decryptWithSecondPassword: async (
          { iterations, password },
          ciphertext
        ) =>
          `decrypted from ${JSON.stringify([
            { ciphertext, iterations, password }
          ])}`
      }

      Wallet.decrypt(securityModule, 'secret', walletSecpass).fork(
        done,
        decrypted => {
          expect(decrypted).toMatchSnapshot()
          done()
        }
      )
    })

    it('should fail when given an incorrect password', done => {
      const securityModule = {
        computeSecondPasswordHash: async () => `wrong hash`,

        decryptWithSecondPassword: async (
          { iterations, password },
          ciphertext
        ) =>
          `decrypted from ${JSON.stringify([
            { ciphertext, iterations, password }
          ])}`
      }

      Wallet.decrypt(securityModule, 'wrong', walletSecpass).fork(error => {
        expect(R.is(Error, error)).toEqual(true)
        expect(error.message).toEqual('INVALID_SECOND_PASSWORD')
        done()
      }, done)
    })
  })

  describe('serializer', () => {
    it('compose(reviver, replacer) should be identity', () => {
      const string = JSON.stringify(wallet)
      const newWallet = JSON.parse(string, serializer.reviver)
      expect(newWallet).toEqual(wallet)
    })
    it('compose(replacer, reviver) should be identity', () => {
      const string = JSON.stringify(wallet)
      const newWallet = JSON.parse(string, serializer.reviver)
      const string2 = JSON.stringify(newWallet)
      expect(string2).toEqual(string)
    })
  })

  // describe('createNew', () => {
  //   const { mnemonic } = walletNewFixture

  //   it('should create a new wallet', () => {
  //     let { guid, sharedKey } = walletNewFixture.wallet
  //     let wallet = Wallet.createNew(guid, sharedKey, mnemonic)
  //     expect(Wallet.toJS(wallet)).toEqual(walletNewFixture.wallet)
  //   })
  // })
})
