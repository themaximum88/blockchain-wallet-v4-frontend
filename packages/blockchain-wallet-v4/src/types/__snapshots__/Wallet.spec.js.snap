// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet decrypt should decrypt 1`] = `
Immutable.Map {
  "addresses": Immutable.Map {
    "19XmKRY66VnUn5irHAafyoTfiwFuGLUxKF": Immutable.Map {
      "addr": "19XmKRY66VnUn5irHAafyoTfiwFuGLUxKF",
      "priv": "decrypted from [{\\"ciphertext\\":\\"ofuAPRjXETbAPACA7qY/wD9aG/xh5XvyZDSxmfqohDdnbFDKXTV87GFfp/55iTU39wXy63qvQP96bdJOBvlAQg==\\",\\"iterations\\":5000,\\"password\\":\\"secret\\"}]",
      "tag": 0,
      "label": "",
      "created_time": 1492721419269,
      "created_device_name": "javascript_web",
      "created_device_version": "3.0",
    },
    "14mQxLtEagsS8gYsdWJbzthFFuPDqDgtxQ": Immutable.Map {
      "addr": "14mQxLtEagsS8gYsdWJbzthFFuPDqDgtxQ",
      "priv": "decrypted from [{\\"ciphertext\\":\\"4MvCwhdK+VxatwL3bbbsrURdjaFisAdUJFfVHiOeEARCrNv2FEevVYcxidNEAuuGpysQzmPME/yQm59+GF7xzg==\\",\\"iterations\\":5000,\\"password\\":\\"secret\\"}]",
      "tag": 0,
      "label": "labeled_imported",
      "created_time": 1492721432222,
      "created_device_name": "javascript_web",
      "created_device_version": "3.0",
    },
    "1JD73aGSdeqKUjJ4ntP4eCyUiuZ3ogJE1u": Immutable.Map {
      "addr": "1JD73aGSdeqKUjJ4ntP4eCyUiuZ3ogJE1u",
      "priv": "decrypted from [{\\"ciphertext\\":null,\\"iterations\\":5000,\\"password\\":\\"secret\\"}]",
      "tag": 0,
      "label": "",
      "created_time": 1492721461228,
      "created_device_name": "javascript_web",
      "created_device_version": "3.0",
    },
  },
  "tx_notes": Immutable.Map {},
  "guid": "50dae286-e42e-4d67-8419-d5dcc563746c",
  "metadataHDNode": "xprv9tygGQP8be7uzNm5Czuy41juTK9pUKnWyZtDxgbmSEcCYa9VdvvtSknEyiKitqqm2TMv14NjXPQ68XLwSdH6Scc5GwXoZ31yRZZysxhVGU7",
  "tx_names": Immutable.List [],
  "double_encryption": false,
  "address_book": Immutable.Map {},
  "hd_wallets": Immutable.List [
    Immutable.Map {
      "seedHex": "decrypted from [{\\"ciphertext\\":\\"YSxSY3q6gfjrmVoRg8GbJU9gyqFFQxitGEmxBKbmE40av+Daa/WIBQf2yguYlrEKPd5O2fgrJVmw9Otf2iVH5w==\\",\\"iterations\\":5000,\\"password\\":\\"secret\\"}]",
      "passphrase": "",
      "mnemonic_verified": false,
      "default_account_idx": 0,
      "accounts": Immutable.List [
        Immutable.Map {
          "label": "My Bitcoin Wallet",
          "archived": false,
          "xpriv": "decrypted from [{\\"ciphertext\\":\\"G1XRCumvdvM10xN1n9ub2JNyk3PphNhoYGQsVKnFi7tKmaDpQVaBmLX4XTWMkFENIGGmKd4PCxwZr6g1RC6RWJFWyXcAs+17DQdv7ZqdiKkj8p0Td3mEM/Ta41KYyxJdjlrT1FyW4HpH0AGejy4rJI43nHlxjpDl0rldP58eaEE=\\",\\"iterations\\":5000,\\"password\\":\\"secret\\"}]",
          "xpub": "xpub6CKNDRQEZmyHUrFdf1HapGEn27ramwpqxZUMEJYUUokoQrz9yLBAiKjGVWDuiCT39udj1r3whqQN89Tar5KrojH8oqSy7ytzJKW8gwmhwD3",
          "address_labels": Immutable.Map {
            "0": Immutable.Map {
              "index": 0,
              "label": "labeled_address",
            },
          },
          "cache": Immutable.Map {
            "receiveAccount": "xpub6F41z8MqNcJMvKQgAd5QE2QYo32cocYigWp1D8726ykMmaMqvtqLkvuL1NqGuUJvU3aWyJaV2J4V6sD7Pv59J3tYGZdYRSx8gU7EG8ZuPSY",
            "changeAccount": "xpub6F41z8MqNcJMwmeUExdCv7UXvYBEgQB29SWq9jyxuZ7WefmSTWcwXB6NRAJkGCkB3L1Eu4ttzWnPVKZ6REissrQ4i6p8gTi9j5YwDLxmZ8p",
          },
          "index": 0,
        },
      ],
    },
  ],
  "sharedKey": "8a260b2b-5257-4357-ac56-7a7efca323ea",
  "options": Immutable.Map {
    "pbkdf2_iterations": 5000,
    "fee_per_kb": 10000,
    "html5_notifications": false,
    "logout_time": 600000,
  },
}
`;

exports[`Wallet encrypt should encrypt 1`] = `
Immutable.Map {
  "addresses": Immutable.Map {
    "19XmKRY66VnUn5irHAafyoTfiwFuGLUxKF": Immutable.Map {
      "addr": "19XmKRY66VnUn5irHAafyoTfiwFuGLUxKF",
      "priv": "encrypted with {\\"iterations\\":5000,\\"password\\":\\"secret\\",\\"plaintext\\":\\"38W3vsxt246Lhawvz81y2HQVnsJrBUX87jSTUUnixPsE\\"}",
      "tag": 0,
      "label": "",
      "created_time": 1492721419269,
      "created_device_name": "javascript_web",
      "created_device_version": "3.0",
    },
    "14mQxLtEagsS8gYsdWJbzthFFuPDqDgtxQ": Immutable.Map {
      "addr": "14mQxLtEagsS8gYsdWJbzthFFuPDqDgtxQ",
      "priv": "encrypted with {\\"iterations\\":5000,\\"password\\":\\"secret\\",\\"plaintext\\":\\"BpD2ZuJjZ8PJPpDX6ZmsKFsXHkL7XV3dt385zghMfF6C\\"}",
      "tag": 0,
      "label": "labeled_imported",
      "created_time": 1492721432222,
      "created_device_name": "javascript_web",
      "created_device_version": "3.0",
    },
    "1JD73aGSdeqKUjJ4ntP4eCyUiuZ3ogJE1u": Immutable.Map {
      "addr": "1JD73aGSdeqKUjJ4ntP4eCyUiuZ3ogJE1u",
      "priv": "encrypted with {\\"iterations\\":5000,\\"password\\":\\"secret\\",\\"plaintext\\":null}",
      "tag": 0,
      "label": "",
      "created_time": 1492721461228,
      "created_device_name": "javascript_web",
      "created_device_version": "3.0",
    },
  },
  "tx_notes": Immutable.Map {},
  "guid": "50dae286-e42e-4d67-8419-d5dcc563746c",
  "dpasswordhash": "hash of {\\"iterations\\":5000,\\"password\\":\\"secret\\"}",
  "metadataHDNode": "xprv9tygGQP8be7uzNm5Czuy41juTK9pUKnWyZtDxgbmSEcCYa9VdvvtSknEyiKitqqm2TMv14NjXPQ68XLwSdH6Scc5GwXoZ31yRZZysxhVGU7",
  "tx_names": Immutable.List [],
  "double_encryption": true,
  "address_book": Immutable.Map {},
  "hd_wallets": Immutable.List [
    Immutable.Map {
      "seedHex": "encrypted with {\\"iterations\\":5000,\\"password\\":\\"secret\\",\\"plaintext\\":\\"6a4d9524d413fdf69ca1b5664d1d6db0\\"}",
      "passphrase": "",
      "mnemonic_verified": false,
      "default_account_idx": 0,
      "accounts": Immutable.List [
        Immutable.Map {
          "label": "My Bitcoin Wallet",
          "archived": false,
          "xpriv": "encrypted with {\\"iterations\\":5000,\\"password\\":\\"secret\\",\\"plaintext\\":\\"xprv9yL1ousLjQQzGNBAYykaT8J3U626NV6zbLYkRv8rvUDpY4f1RnrvAXQneGXC9UNuNvGXX4j6oHBK5KiV2hKevRxY5ntis212oxjEL11ysuG\\"}",
          "xpub": "xpub6CKNDRQEZmyHUrFdf1HapGEn27ramwpqxZUMEJYUUokoQrz9yLBAiKjGVWDuiCT39udj1r3whqQN89Tar5KrojH8oqSy7ytzJKW8gwmhwD3",
          "address_labels": Immutable.Map {
            "0": Immutable.Map {
              "index": 0,
              "label": "labeled_address",
            },
          },
          "cache": Immutable.Map {
            "receiveAccount": "xpub6F41z8MqNcJMvKQgAd5QE2QYo32cocYigWp1D8726ykMmaMqvtqLkvuL1NqGuUJvU3aWyJaV2J4V6sD7Pv59J3tYGZdYRSx8gU7EG8ZuPSY",
            "changeAccount": "xpub6F41z8MqNcJMwmeUExdCv7UXvYBEgQB29SWq9jyxuZ7WefmSTWcwXB6NRAJkGCkB3L1Eu4ttzWnPVKZ6REissrQ4i6p8gTi9j5YwDLxmZ8p",
          },
          "index": 0,
        },
      ],
    },
  ],
  "sharedKey": "8a260b2b-5257-4357-ac56-7a7efca323ea",
  "options": Immutable.Map {
    "pbkdf2_iterations": 5000,
    "fee_per_kb": 10000,
    "html5_notifications": false,
    "logout_time": 600000,
  },
}
`;
